src/Microsoft.VisualStudio.LanguageServices.Razor/TagHelperDescriptorJsonConverter.cs

```diff
@@ -0,0 +1,234 @@
+            var nameComparison = attribute[nameof(RequiredAttributeDescriptor.NameComparison)].Value<int>();
+            var value = attribute[nameof(RequiredAttributeDescriptor.Value)].Value<string>();
+            var valueComparison = attribute[nameof(RequiredAttributeDescriptor.ValueComparison)].Value<int>();
+            var diagnostics = attribute[nameof(RequiredAttributeDescriptor.Diagnostics)].Value<JArray>();
+
+            builder
+                .Name(name)
+                .NameComparisonMode((RequiredAttributeDescriptor.NameComparisonMode)nameComparison)
+                .Value(value)
+                .ValueComparisonMode((RequiredAttributeDescriptor.ValueComparisonMode)valueComparison);
+
+            foreach (var diagnostic in diagnostics)
+            {
+                builder.AddDiagnostic(ReadDiagnostic(diagnostic.Value<JObject>(), serializer));
+            }
+        }
+
+        private void ReadBoundAttribute(ITagHelperBoundAttributeDescriptorBuilder builder, JObject attribute, JsonSerializer serializer)
+        {
+            var descriptorKind = attribute[nameof(BoundAttributeDescriptor.Kind)].Value<string>();
+            if (descriptorKind != ITagHelperBoundAttributeDescriptorBuilder.DescriptorKind)
+            {
+                throw new NotSupportedException();
+            }
+
+            var name = attribute[nameof(BoundAttributeDescriptor.Name)].Value<string>();
+            var typeName = attribute[nameof(BoundAttributeDescriptor.TypeName)].Value<string>();
+            var isEnum = attribute[nameof(BoundAttributeDescriptor.IsEnum)].Value<bool>();
+            var indexerNamePrefix = attribute[nameof(BoundAttributeDescriptor.IndexerNamePrefix)].Value<string>();
+            var indexerTypeName = attribute[nameof(BoundAttributeDescriptor.IndexerTypeName)].Value<string>();
+            var documentation = attribute[nameof(BoundAttributeDescriptor.Documentation)].Value<string>();
+            var diagnostics = attribute[nameof(BoundAttributeDescriptor.Diagnostics)].Value<JArray>();
+            var metadata = attribute[nameof(BoundAttributeDescriptor.Metadata)].Value<JObject>();
+
+            builder
+                .Name(name)
+                .TypeName(typeName)
+                .Documentation(documentation)
+                .AsDictionary(indexerNamePrefix, indexerTypeName);
+
+            if (isEnum)
+            {
+                builder.AsEnum();
+            }
+
+            foreach (var diagnostic in diagnostics)
+            {
+                builder.AddDiagnostic(ReadDiagnostic(diagnostic.Value<JObject>(), serializer));
+            }
+
+            var metadataValue = serializer.Deserialize<Dictionary<string, string>>(metadata.CreateReader());
+            foreach (var item in metadataValue)
+            {
+                if (string.Equals(item.Key, ITagHelperBoundAttributeDescriptorBuilder.PropertyNameKey, StringComparison.Ordinal))
```

Comment:
Can do this outside of the loop and query the metadata. The builder will ensure that duplicates aren't added to the metadata bag.